File: 329-longest-increasing-path-in-a-matrix/longest-increasing-path-in-a-matrix.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        row, col = len(matrix), len(matrix[0])
        
        dp = {}

        def dfs(i, j, prevVal):
            if (i < 0 or j < 0 or i == row or j == col or matrix[i][j] <= prevVal): return 0
            if (i, j) in dp: return dp[(i, j)]
            
            res = 1
            res = max(res, 1 + dfs(i + 1, j, matrix[i][j]))
            res = max(res, 1 + dfs(i - 1, j, matrix[i][j]))
            res = max(res, 1 + dfs(i, j + 1, matrix[i][j]))
            res = max(res, 1 + dfs(i, j - 1, matrix[i][j]))
            dp[(i, j)] = res
            return res
        
        for i in range(row):
            for j in range(col):
                dfs(i, j, -1)
        
        return max(dp.values())

File: 329-longest-increasing-path-in-a-matrix/longest-increasing-path-in-a-matrix.cpp
bool compare(const pair<string, int>& a, const pair<string, int>& b){
    return a.second < b.second;
}
class Solution {
public:
    unordered_map<string, int> cache;
    int row = 0;
    int col = 0;
    
    int dfs(int i, int j, int prevValue, vector<vector<int>>& matrix){
        if (i < 0 || j < 0 || i == row || j == col || matrix[i][j] <= prevValue) return 0;
        string key = to_string(i) + "_" + to_string(j);
        if (cache.contains(key)) return cache[key];
        int res = 1;
        res = max(res, 1 + dfs(i + 1, j, matrix[i][j], matrix));
        res = max(res, 1 + dfs(i - 1, j, matrix[i][j], matrix));
        res = max(res, 1 + dfs(i, j + 1, matrix[i][j], matrix));
        res = max(res, 1 + dfs(i, j - 1, matrix[i][j], matrix));
        cache[key] = res;
        return res;
    }

    int longestIncreasingPath(vector<vector<int>>& matrix) {
        row = matrix.size();
        col = matrix[0].size();
        for (int i = 0; i < row; i++)
            for (int j = 0; j < col; j++)
                dfs(i, j, -1, matrix);
        return max_element(cache.begin(), cache.end(), compare)->second;  
    }
};